src/transcript.rs

```diff
@@ -233,37 +311,47 @@ impl<C: CurveAffine, Type> From<Challenge> for ChallengeScalar<C, Type> {
 
         ChallengeScalar {
             inner: acc,
-            _marker: PhantomData,
+            challenge_type,
         }
     }
 }
 
-impl<C: CurveAffine, Type> ChallengeScalar<C, Type> {
-    /// Obtains a new challenge from the transcript.
-    pub fn get<T: Transcript<C>>(transcript: &mut T) -> Self
-    where
-        C: CurveAffine,
-    {
-        Challenge::get(transcript).into()
+/// The scalar challenge space that samples from the full-width field.
+#[derive(Debug)]
+pub struct ChallengeScalarFull<C: CurveAffine> {
+    _marker: PhantomData<C>,
+}
+
+impl<C: CurveAffine> ChallengeSpace<C> for ChallengeScalarFull<C> {
+    /// This algorithm applies the mapping of Algorithm 1 from the
+    /// [Halo](https://eprint.iacr.org/2019/1021) paper.
+    fn to_challenge_scalar(
+        challenge: Challenge,
+        challenge_type: ChallengeScalarType,
+    ) -> ChallengeScalar<C> {
+        ChallengeScalar {
+            inner: C::Scalar::from_u128(challenge.0),
```

Comment:
This doesn't look full-width. The difference between endoscaling and plain scaling is actually orthogonal to the width. For both, we probably need a width (i.e. entropy input) greater than 128 bits to achieve 128-bit security.